File: FIXES/kron.cc
// This program is in the public domain.

#include <octave/oct.h>

#if !defined( CXX_NEW_FRIEND_TEMPLATE_DECL)
extern void
kron(const Array2<double>&a, const Array2<double>&b, Array2<double>&c);
extern void
kron(const Array2<Complex>&a, const Array2<Complex>&b, Array2<Complex>&c);
#endif

template <class T>
void
kron (const Array2<T>&A, const Array2<T>&B, Array2<T>&C)
{
    C.resize(A.rows()*B.rows(),A.columns()*B.columns());
    int Ar,Ac,Cr,Cc;

    for (Ac = Cc = 0; Ac < A.columns(); Ac++, Cc += B.columns())
      for (Ar = Cr = 0; Ar < A.rows(); Ar++, Cr += B.rows())
	{
	  const T v = A (Ar, Ac);
	  for (int Bc=0; Bc < B.columns(); Bc++)
	    for (int Br=0; Br < B.rows(); Br++)
	      C.xelem(Cr+Br,Cc+Bc) = v*B.elem(Br,Bc);
	}
}

template void
kron(const Array2<double>&a, const Array2<double>&b, Array2<double>&c);
template void
kron(const Array2<Complex>&a, const Array2<Complex>&b, Array2<Complex>&c);

DEFUN_DLD(kron, args,  nargout, "-*- texinfo -*-\n\
@deftypefn {Function File} {} kron (@var{a}, @var{b})\n\
Form the kronecker product of two matrices, defined block by block as\n\
\n\
@example\n\
x = [a(i, j) b]\n\
@end example\n\
\n\
For example,\n\
\n\
@example\n\
@group\n\
kron (1:4, ones (3, 1))\n\
      @result{}  1  2  3  4\n\
          1  2  3  4\n\
          1  2  3  4\n\
@end group\n\
@end example\n\
@end deftypefn")
{
    octave_value_list retval;
    int nargin = args.length();

    if (nargin != 2 || nargout > 1)
    {
	print_usage("kron");
    }
    else if (args(0).is_complex_type() || args(1).is_complex_type())
    {
	ComplexMatrix a (args(0).complex_matrix_value());
	ComplexMatrix b (args(1).complex_matrix_value());
	if (! error_state)
	{
	    ComplexMatrix c;
	    kron (a, b, c);
	    retval (0) = c;
	}
    }
    else
    {
	Matrix a (args(0).matrix_value());
	Matrix b (args(1).matrix_value());
	if (! error_state)
	{
	    Matrix c;
	    kron (a, b, c);
	    retval (0) = c;
	}
    }
    return retval;
}

File: FIXES/kron.m
## Copyright (C) 1996, 1997 John W. Eaton
##
## This file is part of Octave.
##
## Octave is free software; you can redistribute it and/or modify it
## under the terms of the GNU General Public License as published by
## the Free Software Foundation; either version 2, or (at your option)
## any later version.
##
## Octave is distributed in the hope that it will be useful, but
## WITHOUT ANY WARRANTY; without even the implied warranty of
## MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
## General Public License for more details.
##
## You should have received a copy of the GNU General Public License
## along with Octave; see the file COPYING.  If not, write to the Free
## Software Foundation, 59 Temple Place - Suite 330, Boston, MA
## 02111-1307, USA.

## -*- texinfo -*-
## @deftypefn {Function File} {} kron (@var{a}, @var{b})
## Form the kronecker product of two matrices, defined block by block as
##
## @example
## x = [a(i, j) b]
## @end example
##
## For example,
##
## @example
## @group
## kron (1:4, ones (3, 1))
##      @result{}  1  2  3  4
##          1  2  3  4
##          1  2  3  4
## @end group
## @end example
## @end deftypefn

## Author: A. S. Hodel <scotte@eng.auburn.edu>
## Created: August 1993
## Adapted-By: jwe
## Rewritten-By: Paul Kienzle <pkienzle@kienzle.powernet.co.uk>

function x = kron (a, b)
  if (nargin != 2)
    usage ("kron (a, b)");
  endif

  [ra, ca] = size (a);
  [rb, cb] = size (b);
  if (isempty (a) || isempty (b))
    x = zeros (ra*rb, ca*cb);
  else
    rak = 1:ra;    
    rak = rak (ones (1, rb), :);
    cak = 1:ca;    
    cak = cak (ones (1, cb), :);
    rbk = [1:rb]'; 
    rbk = rbk (:, ones (1, ra));
    cbk = [1:cb]'; 
    cbk = cbk (:, ones (1, ca));
    x = a (rak, cak) .* b (rbk, cbk);
  endif

endfunction
